### PR TITLE
SED-3176-implement-k8s-native-plugin-to-run-kw-in-custom-docker-images

### PR DESCRIPTION
**Pull request by jeromecomte**

### CODE REVIEW

File: step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenForecastingContext.java [hunk 1]
```diff
@@ -9,64 +11,104 @@
...
 
-    private String getBestMatchingPool(Map<String, Interest> criteria) {
-        SimpleAffinityEvaluator<Identity, Identity> affinityEvaluator = new SimpleAffinityEvaluator<>();
-        return pools.entrySet().stream()
-                .map(entry -> new Object[]{affinityEvaluator.getAffinityScore(new TokenPretender(Map.of(), criteria), new TokenPretender(entry.getValue(), Map.of())), entry.getKey()})
-                .filter(o -> ((int) o[0]) >= 0).sorted(Comparator.comparingInt(o -> (int) o[0])).map(o -> (String) o[1])
-                .findFirst().orElse(null);
+    private Optional<AgentPoolSpec> getBestMatchingPool(Map<String, Interest> criteria) {
+        PreProvisioningTokenAffinityEvaluator affinityEvaluator = new PreProvisioningTokenAffinityEvaluator();
+        return availableAgentPools.stream()
```
Comment (david-stephan) [inline step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenForecastingContext.java:86]: I got stuck a while to review this piece of code. At least some comment should be added, a refactoring would be nice if possible.

File: step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenForecastingContext.java [hunk 2]
```diff
@@ -9,64 +11,141 @@
...
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
-public class TokenForecastingContext {
+import static step.core.agents.provisioning.AgentPoolProvisioningParameters.supportedParameters;
 
-    protected final Map<String, PoolReservationTracker> poolResourceReservations = new HashMap<>();
-    protected final Map<String, Map<String, String>> pools;
+public class TokenForecastingContext {
+    protected final Map<Key, PoolReservationTracker> poolResourceReservations = new HashMap<>();
```
Comment (david-stephan) [inline step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenForecastingContext.java:23]: PoolReservationTracker being private this field should be private too

**Review by david-stephan — APPROVED** [review]
Approved with minor comments

File: step-core/src/main/java/step/core/agents/provisioning/AgentPoolProvisioningParameter.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    public AgentPoolProvisioningParameter() {
+    }
+
+    /**
+     * @param name the unique name of the provisioning parameter
+     * @param label the label of the provisioning parameter to be displayed to the users
+     * @param tokenSelectionCriteriaToAgentPoolProvisioningParameters defines the consumer that creates the provisioning parameters based on the selection criteria
+     * @param preProvisioningTokenSelectionCriteriaTransformer defines the transformer of token selection criteria.
+     *                                                         Some token attributes are only populated at provisioning time and have to be either removed
```
[the remark this answers is not in the document]
Comment (jeromecomte) [reply step-core/src/main/java/step/core/agents/provisioning/AgentPoolProvisioningParameter.java:27]: This was formatted automatically by intelliJ. It doesn't affect the way it is displayed in the javadoc

**Pull request merged**